### PR TITLE
daemon/config: fix filter type in BuildKit GC config

### LINKED ISSUE #39601: Example Buildkit GC configuration results in daemon error
**Description**

According to [this PR](https://github.com/moby/moby/pull/37846), the correct syntax for buildkit gc configuration is:

```json
{
  "features": {
    "buildkit": true
  },
  "builder": {
    "gc": {
      "enabled": true,
      "policy": [
        {"keepStorage": "10GB", "filter": ["unused-for=2200h"]},
        {"keepStorage": "100GB", "all": true}
      ]
    }
  }
}
```

**Steps to reproduce the issue:**
1. Write the aforementioned config to `/etc/docker/daemon.json`
2. Restart the daemon (e.g. `systemctl restart docker`)

**Describe the results you received:**
Daemon logs the following error and exits with status code 1:

```
unable to configure the Docker daemon with file /etc/docker/daemon.json: json: cannot unmarshal array into Go struct field BuilderGCRule.Filter of type map[string]map[string]bool
```

**Describe the results you expected:**
Daemon should successfully start and properly set the buildkit gc policy

**How to fix:**

```diff
-        {"keepStorage": "10GB", "filter": ["unused-for=2200h"]},
+        {"keepStorage": "10GB", "filter": {"unused-for": {"2200h": true}}},
```

**Output of `dockerd --version`:**
`Docker version 18.09.7, build 2d0083d`

<details>
<summary><strong>Output of <code>docker version</code>:</strong></summary>
<pre>
<code>Client:
 Version:           18.09.7
 API version:       1.39
 Go version:        go1.10.1
 Git commit:        2d0083d
 Built:             Wed Jul  3 12:13:59 2019
 OS/Arch:           linux/amd64
 Experimental:      false
<br>
Server:
 Engine:
  Version:          18.09.7
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.1
  Git commit:       2d0083d
  Built:            Mon Jul  1 19:31:12 2019
  OS/Arch:          linux/amd64
  Experimental:     false</code>
</pre>
</details>

<details>
<summary><strong>Output of <code>docker info</code>:</strong></summary>
<pre>
<code>Containers: xxxxxxx
 Running: xxxxxxx
 Paused: xxxxxxx
 Stopped: xxxxxxx
Images: xxxxxxx
Server Version: 18.09.7
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 
runc version: N/A
init version: v0.18.0 (expected: fec3683b971d9c3ef73f284f176672c44b448662)
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.xxxxxxx
Operating System: Ubuntu 18.04xxxxxxx
OSType: linux
Architecture: x86_64
CPUs: xxxxxxx
Total Memory: xxxxxxx
Name: xxxxxxx
ID: MNWV:EE66:UZOP:DKSM:ZQEZ:J45B:LMSV:WVNF:462I:CLC2:FPQ4:JQRO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
<br>
WARNING: No swap limit support</code>
</pre>
</details>

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Ubuntu 18.04

### PR DESCRIPTION
**Pull request by tiborvass**
Signed-off-by: Tibor Vass <tibor@docker.com>

Fixes #39601

### CODE REVIEW

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by tonistiigi** [comment]
We could possibly use `RawMessage`(for some time) here so we don't break people who used the old format. I guess not a lot of people are affected but if they are it is a failure that causes daemon not to start.

**Comment by tiborvass** [comment]
Ok will update to prevent breaking people

**Comment by thaJeztah** [comment]
> Ok will update to prevent breaking people

Perhaps as a separate commit, and with a `TODO` added to remove the commit after X release(s) ?

**Comment by tiborvass** [comment]
PTAL @tonistiigi

**Comment by tonistiigi** [comment]
@tiborvass CI red

**Comment by tiborvass** [comment]
@tonistiigi updated

**Comment by tiborvass** [comment]
@tonistiigi Updated to your liking, and added another commit for safety. PTAL

File: daemon/config/builder_test.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+func TestBuilderGC(t *testing.T) {
+	tempFile := fs.NewFile(t, "config", fs.WithContent(`{
+  "builder": {
+    "gc": {
+      "enabled": true,
+      "policy": [
+        {"keepStorage": "10GB", "filter": ["unused-for=2200h"]},
+        {"keepStorage": "50GB", "filter": {"unused-for": {"3300h": true}}},
+        {"keepStorage": "100GB", "all": true}
```
Comment (thaJeztah) [inline daemon/config/builder_test.go]: (just a comment); trying to get my head around how the combination of these policies work 🤔😁

Guess I need to read the docs

**Review by tonistiigi — APPROVED** [review]

**Pull request merged**